src/GitHub.InlineReviews/Services/PullRequestSession.cs

```diff
@@ -0,0 +1,276 @@
+        public async Task<IPullRequestSessionFile> GetFile(
+            string relativePath,
+            IEditorContentSource contentSource)
+        {
+            await getFilesLock.WaitAsync();
+
+            try
+            {
+                PullRequestSessionFile file;
+
+                relativePath = relativePath.Replace("\\", "/");
+
+                if (!fileIndex.TryGetValue(relativePath, out file))
+                {
+                    // TODO: Check for binary files.
+                    file = await CreateFile(relativePath, contentSource);
+                    fileIndex.Add(relativePath, file);
+                }
+                else if (contentSource != null && file.ContentSource != contentSource)
+                {
+                    file.ContentSource = contentSource;
+                    await UpdateEditorContent(relativePath);
+                }
+
+                return file;
+            }
+            finally
+            {
+                getFilesLock.Release();
+            }
+        }
+
+        /// <inheritdoc/>
+        public string GetRelativePath(string path)
+        {
+            if (Path.IsPathRooted(path))
+            {
+                var basePath = Repository.LocalPath;
+
+                if (path.StartsWith(basePath) && path.Length > basePath.Length + 1)
+                {
+                    return path.Substring(basePath.Length + 1);
+                }
+            }
+
+            return null;
+        }
+
+        /// <inheritdoc/>
+        public async Task UpdateEditorContent(string relativePath)
+        {
+            PullRequestSessionFile file;
+
+            relativePath = relativePath.Replace("\\", "/");
+
+            if (fileIndex.TryGetValue(relativePath, out file))
+            {
+                var content = await GetFileContent(file);
+
+                file.CommitSha = await CalculateCommitSha(file, content);
+                file.Diff = await service.Diff(Repository, file.BaseSha, relativePath, content);
+
+                foreach (var thread in file.InlineCommentThreads)
+                {
+                    thread.LineNumber = GetUpdatedLineNumber(thread, file.Diff);
+                    thread.IsStale = false;
+                }
+            }
+        }
+
+        public async Task Update(IPullRequestModel pullRequest)
+        {
+            PullRequest = pullRequest;
+
+            foreach (var file in this.fileIndex.Values)
+            {
+                await UpdateFile(file);
+            }
+        }
+
+        async Task UpdateFile(PullRequestSessionFile file)
+        {
+            var content = await GetFileContent(file);
+            var unmodified = await service.IsUnmodifiedAndPushed(Repository, file.RelativePath, content);
+
+            file.BaseSha = PullRequest.Base.Sha;
+            file.CommitSha = await CalculateCommitSha(file, content);
+            file.Diff = await service.Diff(Repository, file.BaseSha, file.RelativePath, content);
```

Comment:
I think we need to calculate the `merge base` here, rather than simply use the `file.BaseSha`. Wonder if we should have an extra `MergeBaseSha` property?